csrc/transform_replay.cpp

```diff
@@ -283,72 +277,81 @@ void TransformReplay::selfReplay(
 
   if (self->hasAllocation()) {
     // Replay producer dimensions.
-    const std::vector<IterDomain*>& self_allocation = self->maybeAllocation();
+    const std::vector<IterDomain*>& self_allocation = self->allocation();
     const std::vector<std::optional<bool>>& self_contiguity =
         self->contiguity();
-    const std::vector<IterDomain*>& self_allocation_no_reduction =
-        TensorDomain::noReductions(self_allocation);
+    NVF_ERROR_EQ(self_allocation.size(), self_contiguity.size());
 
-    // we replay only non-reduction IDs. The reason is that, we might have
-    // non-mapping reduction IDs between self and new_self. This is used in
-    // `RemoveBcastSqueeze`.
-    ReplaySelf replay(self_allocation_no_reduction, axis_map);
     std::vector<IterDomain*> new_alloc_domain;
     std::vector<std::optional<bool>> new_contiguity;
     new_alloc_domain.reserve(self_allocation.size());
-    new_contiguity.reserve(self_allocation.size());
+    new_contiguity.reserve(self_contiguity.size());
 
     // Push back the reduction IDs that are not mapped
-    for (auto id : new_self->logical()) {
-      if (id->isReduction()) {
-        new_alloc_domain.push_back(id);
-        // NOLINTNEXTLINE(modernize-use-emplace)
-        new_contiguity.push_back(std::nullopt);
+    if (ignore_reductions) {
+      for (auto* id : new_self->logical()) {
+        if (id->isReduction()) {
+          new_alloc_domain.push_back(id);
+          // NOLINTNEXTLINE(modernize-use-emplace)
+          new_contiguity.push_back(std::nullopt);
+        }
       }
     }
 
     // Pushing the mapped IDs and corresponding contiguity flags
-    for (size_t i : arange(self_allocation.size())) {
-      IterDomain* id = self_allocation[i];
-      if (id->isReduction()) {
+    ReplaySelf replay(self_allocation, axis_map);
```

Comment:
To check my understanding:
If `self_allocation` has reduction dimensions, but `ignore_reductions=true`, `axis_map` will not contain that id, and hence no replay for it.

Reply:
That's correct. In fact, this version was buggy -- despite of setErrorOnFailure in ReplaySelf::ReplaySelf, `error_on_failure_` wasn't respected in `handle`s. My latest commit fixed that.